File: scala/Practice2/ShortestPath/SpanningTree/Built.ml.scala
// https://atcoder.jp/contests/abc065/tasks/arc076_b

object Main {

  implicit class ScopeFunction[T](x: T) {
    def let[S](f: T => S): S = f(x)
  }

  val readInput = () =>
    io.StdIn.readLine.toInt.let(n =>
      Array.range(0, n).map(i =>
        io.StdIn.readLine
          .split(" ")
          .map(_.toInt)
          .let(x => (i, x(0), x(1)))
      )
    )

  val makeEdges = (vs: Array[Tuple3[Int, Int, Int]]) =>
    (
      vs.sortWith((a, b) => if (a._2 != b._2) a._2 < b._2 else a._3 < b._3).let(vsX =>
        List.range(1, vs.length).map(i => (vsX(i - 1)._1, vsX(i)._1, vsX(i)._2 - vsX(i - 1)._2))
      ),
      vs.sortWith((a, b) => if (a._3 != b._3) a._3 < b._3 else a._2 < b._2).let(vsY =>
        List.range(1, vs.length).map(i => (vsY(i - 1)._1, vsY(i)._1, vsY(i)._3 - vsY(i - 1)._3))
      )
    )
    .let { case (a, b) => a ++ b }

  val findTree = (table: Array[Int], x: Int) =>
    Array(true).let(next =>
      Stream.continually(0)
        .takeWhile(_x => next(0))
        .scanLeft(x)((current, _x) =>
          table(current).let(parent =>
            (if (parent == current) next(0) = false).let(_x => parent)
          )
        )
        .reverse
        .toList
    )

  val findRoot = (table: Array[Int], x: Int) =>
    findTree(table, x).let(tree =>
      tree.tail
        .foreach(item => table(item) = tree.head)
        .let(_x => tree.head)
    )

  val updateRoot = (table: Array[Int], x: Int, newRoot: Int) =>
    findTree(table, x).let(tree =>
      tree.head.let(oldRoot =>
        tree
          .foreach(item => table(item) = newRoot)
          .let(_x => oldRoot)
      )
    )

  val kruskal = (v: Int, edges: List[Tuple3[Int, Int, Int]]) =>
    Array.range(0, v).let(table =>
      edges
        .sortBy(_._3)
        .foldLeft(0L) { case (cost, (a, b, c)) =>
          findRoot(table, a).let(rA =>
            updateRoot(table, b, rA).let(rB =>
              if (rA != rB)
                cost + c.toLong
              else
                cost
            )
          )
        }
    )

  def main(args: Array[String]) = (
    readInput()
      .let(vs => (vs.length, makeEdges(vs)))
      .let(kruskal.tupled)
      .let(println)
  )
}

// https://atcoder.jp/contests/abc065/submissions/24155699

File: fs/Practice2/SpanningTree/Built.ml.fs
// https://atcoder.jp/contests/abc065/tasks/arc076_b

let readInput = fun () ->
  stdin.ReadLine()
  |> int
  |> fun n ->
    seq { 1..n }
    |> Seq.map(fun i ->
      stdin.ReadLine().Split()
      |> Array.map int
      |> fun x -> (i, x.[0], x.[1])
    )
    |> Seq.toArray

let makeEdges = fun (vs: (int * int * int) []) ->
  [
    (
      vs
      |> Array.sortWith (fun (_, x1, y1) (_, x2, y2) -> if x1 <> x2 then x1 - x2 else y1 - y2)
      |> fun vsX ->
        [1..vs.Length - 1]
        |> List.map (fun i -> (vsX.[i - 1], vsX.[i]) |> fun ((i1, x1, _), (i2, x2, _)) -> (i1, i2, x2 - x1))
    );
    (
      vs
      |> Array.sortWith (fun (_, x1, y1) (_, x2, y2) -> if y1 <> y2 then y1 - y2 else x1 - x2)
      |> fun vsY ->
        [1..vs.Length - 1]
        |> List.map (fun i -> (vsY.[i - 1], vsY.[i]) |> fun ((i1, _, y1), (i2, _, y2)) -> (i1, i2, y2 - y1))
    );
  ]
  |> List.concat

let findTree = fun (table: int []) (x: int) ->
  ([| true |])
  |> fun next ->
    Seq.initInfinite ignore
    |> Seq.takeWhile (fun () -> next.[0])
    |> Seq.scan (
      fun current _ ->
        table.[current]
        |> fun parent ->
          if parent = current
          then (next.[0] <- false) |> fun _ -> current
          else parent
    ) x
    |> Seq.rev
    |> Seq.toList

let findRoot = fun (table: int []) ->
  (findTree table) >> (
    fun tree ->
      tree
      |> List.tail
      |> List.iter (fun x -> table.[x] <- tree |> Seq.head)
      |> fun () -> tree |> Seq.head
  )

let updateRoot = fun (table: int []) (newRoot: int) ->
  (findTree table) >> (
    fun tree ->
      tree
      |> Seq.head
      |> fun oldRoot ->
        tree
        |> List.iter (fun x -> table.[x] <- newRoot)
        |> fun () -> oldRoot
  )

let kruskal = fun (v: int, edges: seq<int * int * int>) ->
  [| 0..v |]
  |> fun table ->
    edges
    |> Seq.sortBy(fun (_, _, c) -> c)
    |> Seq.fold (fun cost (a, b, c) ->
      (findRoot table a)
      |> fun rA ->
        (updateRoot table rA b)
        |> fun rB ->
          if (rA <> rB)
          then cost + int64 c
          else cost
    ) 0L

()
|> readInput
|> fun vs -> (vs.Length, makeEdges vs)
|> kruskal
|> printfn "%d"

// https://atcoder.jp/contests/abc065/submissions/24159032
